src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -207,21 +207,82 @@ impl LocationState {
     }
 
     /// Records a new access, so that future access can potentially be skipped
-    /// by `skip_if_known_noop`.
-    /// The invariants for this function are closely coupled to the function above:
-    /// It MUST be called on child accesses, and on foreign accesses MUST be called
-    /// when `skip_if_know_noop` returns `Recurse`, and MUST NOT be called otherwise.
-    /// FIXME: This optimization is wrong, and is currently disabled (by ignoring the
-    /// result returned here). Since we presumably want an optimization like this,
-    /// we should add it back. See #3864 for more information.
-    #[allow(unused)]
+    /// by `skip_if_known_noop`. It needs to be called after some accesses, in order
+    /// to record this new access.
+    /// The rules for when it needs to be called are tightly coupled to `skip_if_known_noop` above:
+    /// If `skip_if_known_noop` says that a node should be visited, this function MUST be called afterwards.
+    /// If `skip_if_known_noop` says that a node (and its subtree) should be skipped, this function
+    ///   MUST NOT be called for this node (nor any of its children).
     fn record_new_access(&mut self, access_kind: AccessKind, rel_pos: AccessRelatedness) {
+        debug_assert!(matches!(
+            self.skip_if_known_noop(access_kind, rel_pos),
+            ContinueTraversal::Recurse
+        ));
         if rel_pos.is_foreign() {
             self.latest_foreign_access = Some(access_kind);
         } else {
             self.latest_foreign_access = None;
         }
     }
+
+    /// Check if the latest recorded foreign update needs to be updated after a retag.
+    /// Retags can violate the invariant on `LocationState::latest_foreign_access`,
+    /// which states that a node's children survive accesses at least as strong as that node itself.
+    /// While retags act as a read for the offsets specified in the retag, it is also present at the other
+    /// offsets, but marked as "lazy." Such lazy nodes can, however, still be affected by foreign accesses.
+    /// This would violate the invariants that children always survive accesses at least as strong as the last
+    /// recorded foreign access at the parent.
+    /// To restore the invariant, we need to weaken the information stored in the parent, until it is weak enough
+    /// to encompass the newly added lazy node.
+    /// This function compares two accesses, and indicates if the parents needs to be updated.
+    /// It returns `true` iff `last_recorded` is stronger than `happening_now`, where
+    /// `None` is weakest and `Some(Write)` is strongest.
+    pub fn foreign_access_requires_update(
+        last_recorded: Option<AccessKind>,
+        happening_now: Option<AccessKind>,
```

Comment:
`happening_now` is the strongest access that the new node would survive, isn't it? I am confused by the name.

Reply:
`happening_now` is what we are weakening everything to. I agree the name is confusing. How about `at_most` instead?

Reply:
> How about at_most instead?

Is this entire logic just doing `self.latest_foreign_access.min(strongest_survivable)`? Maybe we should use something more like that.